app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.formtemplate
+
+import uk.gov.hmrc.gform.core.{ Invalid, Valid, ValidationResult }
+import uk.gov.hmrc.gform.core.ValidationResult.BooleanToValidationResultSyntax
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+
+class BooleanExprValidator(
+  indexLookup: Map[FormComponentId, Int],
+  wrapperType: BooleanExprWrapperType
+)(
+  idx: Int
+) {
+
+  private def extractFcIds(expr: Expr): List[FormComponentId] = expr match {
+    case Add(left, right)         => extractFcIds(left) ::: extractFcIds(right)
+    case Subtraction(left, right) => extractFcIds(left) ::: extractFcIds(right)
+    case Multiply(left, right)    => extractFcIds(left) ::: extractFcIds(right)
+    case Sum(field1)              => extractFcIds(field1)
+    case FormCtx(fcId)            => List(fcId)
+    case _                        => Nil
+  }
+
+  private def extractFcIds(expr: DateExpr): List[FormComponentId] = expr match {
+    case DateExprWithOffset(dateExpr, _, _)       => extractFcIds(dateExpr)
+    case DateFormCtxVar(FormCtx(formComponentId)) => List(formComponentId)
+    case _                                        => Nil
+  }
+
+  private def validateIdIdx(id: FormComponentId) =
+    indexLookup
+      .get(id)
+      .map { idIdx =>
+        wrapperType
+          .isForwardReference(idIdx, idx)
+          .validationResult(s"id '${id.value}' named in $wrapperType is forward reference, which is not permitted")
+      }
+      .getOrElse(Invalid(s"id '${id.value}' named in $wrapperType expression does not exist in a form"))
```

Comment:
Minor: "id '${id.value}' named in $wrapperType expression does not exist in the current form"

Reply:
sure